README.md
# react-leaflet-vectorgrid

[![version](https://img.shields.io/npm/v/react-leaflet-vectorgrid.svg?style=plastic)](http://npm.im/react-leaflet-vectorgrid)
[![react-leaflet compatibility](https://img.shields.io/npm/dependency-version/react-leaflet-vectorgrid/peer/react-leaflet.svg?style=plastic)](https://github.com/mhasbie/react-leaflet-vectorgrid)
[![travis build](https://img.shields.io/travis/mhasbie/react-leaflet-vectorgrid.svg?style=plastic)](https://travis-ci.org/mhasbie/react-leaflet-vectorgrid)
[![dependencies](https://img.shields.io/david/mhasbie/react-leaflet-vectorgrid.svg?style=plastic)](https://david-dm.org/mhasbie/react-leaflet-vectorgrid)
[![peer dependencies](https://img.shields.io/david/peer/mhasbie/react-leaflet-vectorgrid.svg?style=plastic)](https://david-dm.org/mhasbie/react-leaflet-vectorgrid?type=peer)
[![issues](https://img.shields.io/github/issues/mhasbie/react-leaflet-vectorgrid.svg?style=plastic)](https://github.com/mhasbie/react-leaflet-vectorgrid/issues)
[![downloads](https://img.shields.io/npm/dt/react-leaflet-vectorgrid.svg?style=plastic)](http://npm-stat.com/charts.html?package=react-leaflet-vectorgrid&from=2018-01-01)
[![MIT License](https://img.shields.io/npm/l/react-leaflet-vectorgrid.svg?style=plastic)](http://opensource.org/licenses/MIT)


React wrapper of [Leaflet.VectorGrid](https://github.com/Leaflet/Leaflet.VectorGrid)
for [react-leaflet](https://github.com/PaulLeCam/react-leaflet).

Display gridded vector data (sliced [GeoJSON](http://geojson.org/), [TopoJSON](https://github.com/mbostock/topojson/wiki) or [protobuf vector tiles](https://github.com/mapbox/vector-tile-spec)) in [Leaflet](http://www.leafletjs.com) 1.0.0

*Tested with Leaflet 1.3.4 and React-Leaflet 1.9.1, React-Leaflet 2.1.4*

## Demos

| Version	| Demo	| Description			|
| ---		| ---	| ---					|
| react-leaflet@1.9.1 | [`JSFiddle`](https://jsfiddle.net/m_hasbie/rnx44us3/), [`CodePen`](https://codepen.io/m_hasbie/full/jzpbwj/) | Sliced GeoJSON |
| react-leaflet@2.1.4 | [`JSFiddle`](http://jsfiddle.net/m_hasbie/Lkp96dj3/), [`CodePen`](https://codepen.io/m_hasbie/full/jvgyPq/) | Sliced GeoJSON |



## Installation

### Install via NPM

```bash
npm install --save react-leaflet-vectorgrid
```

`react-leaflet-vectorgrid` requires `lodash` as [`peerDependency`](https://docs.npmjs.com/files/package.json#peerdependencies)

(React, PropTypes, Leaflet, react-leaflet also should be installed)
```bash
npm install --save lodash
```

## Usage example

### Slicer

```javascript
import { Map, TileLayer } from 'react-leaflet';
import VectorGrid from 'react-leaflet-vectorgrid';

const options = {
	type: 'slicer',
	data: {geojson},
	idField: 'OBJECTID',
	tooltip: 'NAME',
	popup: (layer) => `<div>${layer.properties.NAME}</div>`,
	style: {
		weight: 0.5,
		opacity: 1,
		color: '#ccc',
		fillColor: '#390870',
		fillOpacity: 0.6,
		fill: true,
		stroke: true
	},
	hoverStyle: {
		fillColor: '#390870',
		fillOpacity: 1
	},
	activeStyle: {
		fillColor: '#390870',
		fillOpacity: 1
	},
	zIndex: 401
};

<Map center={[2.935403, 101.448205]} zoom={4}>
  <TileLayer
    url="https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png"
    attribution='&copy; <a href="http://osm.org/copyright">OpenStreetMap</a> contributors'
  />

  <VectorGrid {...options} onClick={this.onClick} />
</Map>
```

#### Options

Option          | Type      | Default | Description
--------------- | --------- | ------- | -------------
`data`          | `Object`  | `{}`    | Required when using type `slicer`. A valid [GeoJSON FeatureCollection object](http://geojson.org/geojson-spec.html).
`type`          | `String`  | `'slicer'`| Decides between using [VectoGrid.Slicer](http://leaflet.github.io/Leaflet.VectorGrid/vectorgrid-api-docs.html#vectorgrid-slicer) and [VectorGrid.Protobuf](http://leaflet.github.io/Leaflet.VectorGrid/vectorgrid-api-docs.html#vectorgrid-protobuf). Available options: `slicer`, `protobuf`.
`idField`       | `String`   | `''`   | A unique identifier field in the vector feature.
`tooltip`       | `String` | `function`     | `undefined`   | Show tooltip on vector features. Set to feature properties name to use that properties value as tooltip. Or pass a function that will return a string. e.g. `function(feature) { return feature.properties.NAME; }`
`popup`   | `Function`   | `undefined`    | Similar to `tooltip`, this props will be passed to leaflet `bindPopup` function to create popup for vector features.
`style`   | `Object`  	 | `undefined`    | Apply default style to all vector features. Use this props when not using `vectorTileLayerStyles`
`hoverStyle`   | `Object`  	 | `undefined`    | Style to apply to features on `mouseover` event.
`activeStyle`   | `Object`  	 | `undefined`    | Style to apply to features on `click` event. Can be use to show user selection when feature is clicked. Double click to clear selection.
`zIndex`   | `Integer`   | `undefined`    | Sets the `VectorGrid` z-index.
`interactive`   | `Boolean`   | `true`    | Whether `VectorGrid` fires `Interactive` Layer events.
`onClick` |  `Function`     | `undefined`    | Listens to `VectorGrid` `click` events. `interactive` option must be set to `true`.
`onMouseover` |  `Function`     | `undefined`    | Listens to `VectorGrid` `mouseover` events. `interactive` option must be set to `true`.
`onMouseout` |  `Function`     | `undefined`    | Listens to `VectorGrid` `mouseout` events. `interactive` option must be set to `true`.
`onDblclick` |  `Function`     | `undefined`    | Listens to `VectorGrid` `dblclick` events. `interactive` option must be set to `true`.

### Protobuf

```javascript
import { Map, TileLayer } from 'react-leaflet';
import VectorGrid from 'react-leaflet-vectorgrid';

const options = {
	type: 'protobuf',
	url: 'https://free-{s}.tilehosting.com/data/v3/{z}/{x}/{y}.pbf.pict?key={key}'
	vectorTileLayerStyles: { ... },
    subdomains: 'abcd',
    key: 'abcdefghi01234567890'
};

<Map center={[2.935403, 101.448205]} zoom={4}>
  <TileLayer
    url="https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png"
    attribution='&copy; <a href="http://osm.org/copyright">OpenStreetMap</a> contributors'
  />

  <VectorGrid {...options} />
</Map>
```

#### Options

Option          | Type      | Default | Description
--------------- | --------- | ------- | -------------
`type`          | `String`  | `'slicer'`| Decides between using [VectoGrid.Slicer](http://leaflet.github.io/Leaflet.VectorGrid/vectorgrid-api-docs.html#vectorgrid-slicer) and [VectorGrid.Protobuf](http://leaflet.github.io/Leaflet.VectorGrid/vectorgrid-api-docs.html#vectorgrid-protobuf). Available options: `slicer`, `protobuf`.
`url`           | `String`  | `''`    | Required when using type `protobuf`. Pass a url template that points to vector tiles (usually `.pbf` or `.mvt`).
`subdomains`    | `String`  | `'abc'` | Akin to the `subdomains` option to `L.TileLayer`.
`accessKey`  			| `String`  | `''`    | Tile server access key.
`accessToken`  		| `String`  | `''`    | Tile server access token.
`vectorTileLayerStyles`   | `Object`  	 | `undefined`    | A data structure holding initial symbolizer definitions for the vector features. Refer [Leaflet.VectorGrid doc](http://leaflet.github.io/Leaflet.VectorGrid/vectorgrid-api-docs.html#styling-vectorgrids) for more info.

### Usage with React-Leaflet v2

This is compatible with version 2 of React-Leaflet, but you have to wrap the `VectorGrid` using the [`withLeaflet` higher-order component](https://react-leaflet.js.org/docs/en/context.html) to give it access to the new context mechanism. For example:

```javascript
import { Map, withLeaflet } from 'react-leaflet';
import VectorGridDefault from 'react-leaflet-vectorgrid';

// wrap the VectorGrid component using `withLeaflet` HOC
const VectorGrid = withLeaflet(VectorGridDefault);

// use <VectorGrid /> component as you would use it in react-leaflet v1

```

# Credits
Credits goes to all the [contributors of the original work](https://github.com/Leaflet/Leaflet.VectorGrid/graphs/contributors) and [everyone who contributed to this project](https://github.com/mhasbie/react-leaflet-vectorgrid/graphs/contributors).

# License

MIT License


CHANGELOG.md
react-leaflet-vectorgrid Changelog
=========================


## 2.2.0 (2019-02-01)

* Update dependencies
* Update README
* Add example
* Add unit test
* Fix bugs


__tests__/stubs.js
class Worker {
	constructor(stringUrl) {
		this.url = stringUrl;
		this.onmessage = () => {};
	}

	postMessage(msg) {
		this.onmessage(msg);
	}
}

window.Worker = Worker;


__tests__/enzyme.js
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

// Configure Enzyme for the appropriate React adapter
Enzyme.configure({ adapter: new Adapter() });

// Re-export all enzyme exports
export * from 'enzyme';


webpack.config.js
/* eslint-disable */
module.exports = {
	entry: './dist/react-leaflet-vectorgrid.min.js',
	output: {
		library: {
			root: 'ReactLeafletVectorGrid',
			amd: 'react-leaflet-vectorgrid',
			commonjs: 'react-leaflet-vectorgrid'
		},
		libraryExport: 'default',
		libraryTarget: 'umd'
	},
	externals: {
		leaflet: {
			commonjs: 'leaflet',
			commonjs2: 'leaflet',
			root: 'L'
		},
		'react-leaflet': {
			commonjs: 'react-leaflet',
			commonjs2: 'react-leaflet',
			root: 'ReactLeaflet'
		},
		react: {
			commonjs: 'react',
			commonjs2: 'react',
			root: 'React'
		}
	},
	mode: 'production',
	module: {
		rules: [
			{
				test: /\.(js|jsx)$/,
                exclude: /(node_modules|bower_components)/,
                loader: 'babel-loader',
                options: {
                    presets: ['env', 'react'],
                    plugins: ['transform-class-properties', 'transform-object-rest-spread']
                }
			},
			{
				test: /\.css$/,
				exclude: /node_modules/,
				use: [
					{ loader: 'style-loader' },
					{ loader: 'css-loader' }
				]
			}
		]
	}
};


src/VectorGrid.js
import { MapLayer } from 'react-leaflet';
import L from 'leaflet';
import 'leaflet.vectorgrid';
import isObject from 'lodash/isObject';
import isFunction from 'lodash/isFunction';
import isString from 'lodash/isString';
import isEmpty from 'lodash/isEmpty';
import clone from 'lodash/clone';
import cloneDeep from 'lodash/cloneDeep';
import extend from 'lodash/extend';
import merge from 'lodash/merge';
import has from 'lodash/has';
import find from 'lodash/find';

function patchVectorGridLayer(obj) {
	// Fix error for point data.
	// eg. mouseover does not work without this.
	obj._createLayer_orig = obj._createLayer;
	obj._createLayer = function (feat, pxPerExtent, layerStyle) {
		let layer = this._createLayer_orig(feat, pxPerExtent, layerStyle);
		if (feat.type === 1) {
			layer.getLatLng = null;
		}
		return layer;
	};

	// do this for chaining
	return obj;
}

export default class VectorGrid extends MapLayer {
	createLeafletElement(props) {
		const {
			map,
			pane,
			layerContainer
		} = props.leaflet || this.context;
		const {
			data,
			zIndex,
			style,
			hoverStyle,
			activeStyle,
			onClick,
			onMouseover,
			onMouseout,
			onDblclick,
			onContextmenu,
			vectorTileLayerStyles,
			url,
			maxNativeZoom,
			maxZoom,
			minZoom,
			subdomains,
			accessKey,
			accessToken,
			type = 'slicer',
			interactive = true,
			...rest
		} = props;
		
		// `leaflet` prop is being passed down by withLeaflet() HOC wrapper in react-leaflet v2
		// It caused the plugin to break. We don't need to pass it down to L.vectorgrid
		delete(rest.leaflet);

		// get feature base styling
		const baseStyle = (properties, zoom) => {
			if (isFunction(style)) {
				return style(properties);
			} else if (isObject(style)) {
				return style;
			}
			return {
				weight: 0.5,
				opacity: 1,
				color: '#ccc',
				fillColor: '#390870',
				fillOpacity: 0.6,
				fill: true,
				stroke: true
			};
		};
		this.highlight = null;
		this.active = null;

		let vectorGrid;

		if (type === 'protobuf') {
			vectorGrid = L.vectorGrid.protobuf(url, {
				vectorTileLayerStyles,
				interactive,
				url,
				maxNativeZoom,
				subdomains,
				key: accessKey,
				token: accessToken,
				zIndex: zIndex || Number(layerContainer._panes[pane] && layerContainer._panes[pane].style.zIndex),
				getFeatureId: feature => this._getFeatureId(feature),
				rendererFactory: L.svg.tile,
				maxZoom: maxZoom || map.getMaxZoom(),
				minZoom: minZoom || map.getMinZoom(),
				// pass through other props so we can change lower level options of VectorGrid
				...rest
			});
		} else {
			vectorGrid = L.vectorGrid.slicer(data, {
				interactive,
				zIndex: zIndex || Number(layerContainer._panes[pane] && layerContainer._panes[pane].style.zIndex),
				getFeatureId: feature => this._getFeatureId(feature),
				rendererFactory: L.svg.tile,
				maxZoom: maxZoom || map.getMaxZoom(),
				minZoom: minZoom || map.getMinZoom(),
				vectorTileLayerStyles: vectorTileLayerStyles || {
					sliced: (properties, zoom) => {
						const bs = baseStyle(properties, zoom);
						bs.fill = true;
						bs.stroke = true;
						return bs;
					}
				},
				// pass through other props so we can change lower level options of VectorGrid
				...rest
			});
		}

		return patchVectorGridLayer(vectorGrid)
			.on('mouseover', (e) => {
				const {
					properties
				} = e.layer;
				this._propagateEvent(onMouseover, e);

				// on mouseover styling
				let st;
				const featureId = this._getFeatureId(e.layer);
				if (isFunction(hoverStyle)) {
					st = hoverStyle(properties);
				} else if (isObject(hoverStyle)) {
					st = cloneDeep(hoverStyle);
				}
				if (!isEmpty(st) && featureId) {
					this.clearHighlight();
					this.highlight = featureId;
					const base = cloneDeep(baseStyle(properties));
					const hover = extend(base, st);
					this.setFeatureStyle(featureId, hover);
				}
			})
			.on('mouseout', (e) => {
				this._propagateEvent(onMouseout, e);
				this.clearHighlight();
			})
			.on('click', (e) => {
				const {
					properties
				} = e.layer;
				const featureId = this._getFeatureId(e.layer);

				this._propagateEvent(onClick, e);

				// set active style
				let st;
				if (isFunction(activeStyle)) {
					st = activeStyle(properties);
				} else if (isObject(activeStyle)) {
					st = cloneDeep(activeStyle);
				}
				if (!isEmpty(st) && featureId) {
					this.clearActive();
					this.active = featureId;
					const base = cloneDeep(baseStyle(properties));
					const active = extend(base, st);
					this.setFeatureStyle(featureId, active);
				}
			})
			.on('dblclick', (e) => {
				this._propagateEvent(onDblclick, e);
				this.clearActive();
			})
			.on('contextmenu', (e) => {
				this._propagateEvent(onContextmenu, e);
				this.clearActive();
			});
	}

	componentDidMount() {
		const {
			layerContainer
		} = this.props.leaflet || this.context;
		const {
			tooltipClassName = '',
			tooltip = null,
			popup = null
		} = this.props;
		this.leafletElement.addTo(layerContainer);
		// bind tooltip
		if (tooltip) {
			this.leafletElement.bindTooltip((layer) => {
				if (isFunction(tooltip)) {
					return tooltip(layer);
				} else if (isString(tooltip) && has(layer.properties, tooltip)) {
					return String(layer.properties[tooltip]);
				} else if (isString(tooltip)) {
					return tooltip;
				}
				return '';
			}, {
				sticky: true,
				direction: 'auto',
				className: tooltipClassName
			});
		}
		// bind popup
		// don't need all that extra logic, just make popup match the vector grid options.
		// TODO add the ability to pass through popup options
		if (popup) {
			// create a popup and bind it?
			// const newPopup = L.popup().setContent(popup);
			this.leafletElement.bindPopup(popup);
		}
	}

	_getFeatureId(feature) {
		const {
			idField
		} = this.props;
		if (isFunction(idField)) {
			return idField(feature);
		} else if (isString(idField)) {
			return feature.properties[idField];
		}
	}

	_propagateEvent(eventHandler, e) {
		if (!isFunction(eventHandler)) return;
		const featureId = this._getFeatureId(e.layer);
		const feature = this.getFeature(featureId);
		const event = cloneDeep(e);
		const mergedEvent = merge(event.target, {
			feature
		});
		eventHandler(event);
	}

	setFeatureStyle(id, style) {
		this.leafletElement.setFeatureStyle(id, style);
	}

	resetFeatureStyle(id) {
		this.leafletElement.resetFeatureStyle(id);
	}

	clearHighlight() {
		if (this.highlight && this.highlight !== this.active) {
			this.resetFeatureStyle(this.highlight);
		}
		this.highlight = null;
	}

	clearActive() {
		if (this.active) {
			this.resetFeatureStyle(this.active);
		}
		this.active = null;
	}

	getFeature(featureId) {
		const {
			data,
			idField
		} = this.props;
		if (isEmpty(data) || isEmpty(data.features)) return {};
		const feature = find(data.features, ({
			properties
		}) => properties[idField] === featureId);
		return cloneDeep(feature);
	}
}